pkg/services/ngalert/sender/router.go

```diff
@@ -0,0 +1,268 @@
+			if err != nil {
+				d.logger.Error("failed to apply configuration", "err", err, "org", cfg.OrgID)
+				continue
+			}
+			d.SendersCfgHash[cfg.OrgID] = cfg.AsSHA256()
+			continue
+		}
+
+		// No sender and have Alertmanager(s) to send to - start a new one.
+		d.logger.Info("creating new sender for the external alertmanagers", "org", cfg.OrgID, "alertmanagers", cfg.Alertmanagers)
+		s, err := New()
+		if err != nil {
+			d.logger.Error("unable to start the sender", "err", err, "org", cfg.OrgID)
+			continue
+		}
+
+		d.Senders[cfg.OrgID] = s
+		s.Run()
+
+		err = s.ApplyConfig(cfg)
+		if err != nil {
+			d.logger.Error("failed to apply configuration", "err", err, "org", cfg.OrgID)
+			continue
+		}
+
+		d.SendersCfgHash[cfg.OrgID] = cfg.AsSHA256()
+	}
+
+	sendersToStop := map[int64]*Sender{}
+
+	for orgID, s := range d.Senders {
+		if _, exists := orgsFound[orgID]; !exists {
+			sendersToStop[orgID] = s
+			delete(d.Senders, orgID)
+			delete(d.SendersCfgHash, orgID)
+		}
+	}
+	d.AdminConfigMtx.Unlock()
+
+	// We can now stop these senders w/o having to hold a lock.
+	for orgID, s := range sendersToStop {
+		d.logger.Info("stopping sender", "org", orgID)
+		s.Stop()
+		d.logger.Info("stopped sender", "org", orgID)
+	}
+
+	d.logger.Debug("finish of admin configuration sync")
+
+	return nil
+}
+
+func (d *AlertsRouter) Send(key models.AlertRuleKey, alerts definitions.PostableAlerts) error {
+	logger := d.logger.New("rule_uid", key.UID, "org", key.OrgID)
+	if len(alerts.PostableAlerts) == 0 {
+		logger.Debug("no alerts to notify about")
+		return nil
+	}
+	// Send alerts to local notifier if they need to be handled internally
+	// or if no external AMs have been discovered yet.
+	var localNotifierExist, externalNotifierExist bool
+	if d.SendAlertsTo[key.OrgID] == models.ExternalAlertmanagers && len(d.AlertmanagersFor(key.OrgID)) > 0 {
+		logger.Debug("no alerts to put in the notifier")
+	} else {
+		logger.Debug("sending alerts to local notifier", "count", len(alerts.PostableAlerts), "alerts", alerts.PostableAlerts)
+		n, err := d.MultiOrgNotifier.AlertmanagerFor(key.OrgID)
+		if err == nil {
+			localNotifierExist = true
+			if err := n.PutAlerts(alerts); err != nil {
+				logger.Error("failed to put alerts in the local notifier", "count", len(alerts.PostableAlerts), "err", err)
+			}
+		} else {
+			if errors.Is(err, notifier.ErrNoAlertmanagerForOrg) {
+				logger.Debug("local notifier was not found")
+			} else {
+				logger.Error("local notifier is not available", "err", err)
+			}
+		}
+	}
+
+	// Send alerts to external Alertmanager(s) if we have a sender for this organization
+	// and alerts are not being handled just internally.
+	d.AdminConfigMtx.RLock()
+	defer d.AdminConfigMtx.RUnlock()
+	s, ok := d.Senders[key.OrgID]
+	if ok && d.SendAlertsTo[key.OrgID] != models.InternalAlertmanager {
+		logger.Debug("sending alerts to external notifier", "count", len(alerts.PostableAlerts), "alerts", alerts.PostableAlerts)
+		s.SendAlerts(alerts)
```

Comment:
I looked at the code too and unless I have also misunderstood I think Josh is correct?

Reply:
On a second read after Josh's comment, I agree. I got crosseyed somewhere, because I ended up reading this code:
https://github.com/prometheus/prometheus/blob/44fcf876caad9a5d28b92a728d2c98c34015d377/notifier/notifier.go#L324